MPS notes in contributing.md (#2475)
pytorch/torchtune: CONTRIBUTING.md

pip install -e ".[dev]"
```

> &nbsp;

Reviewer: I have two points here:

1) This should be inside a `> [!NOTE]` tag
2) These instructions should be be general to other non-cuda hardware, and just link to the bitsandbytes docs page for installation on other platforms, rather than providing an explicit install command.

&nbsp;

> [!NOTE]
> ## Installation on Non-CUDA Platforms

> > All steps remain the same except for **Step 2**. [bitsandbytes>=0.43.0](https://github.com/bitsandbytes-foundation/bitsandbytes.git) is not fully supported on non-CUDA platforms. Therefore, you need to remove it from `pyproject.toml` and install it separately by following the instructions provided in the [official documentation](https://huggingface.co/docs/bitsandbytes/main/en/installation).

Reviewer: ```suggestion
> PyPI releases for `bitsandbytes` may not be available for platforms without CUDA support, or which use other accelerator backends (e.g. MPS or XPU). If this is the case, you must remove the `bitsandbytes` dependency in the [`project.toml`](pyproject.toml) file. You may also wish to manually install `bitsandbytes` - please see the [`bitsandbytes` multi-backend documentation](https://huggingface.co/docs/bitsandbytes/main/en/installation#multi-backend) for more information.
```

Note that the above flags can be combined with other pytest flags, so e.g. `pytest tests -m integration_test -k 'test_loss'` will run only recipe tests matching the substring `test_loss`.

> > [!NOTE]
> > It is important to note that reference loss values have been calculated on CPU/GPU. As a result, some tests may be skipped on non-CUDA platforms. Keep this in mind when adding new tests involving calculations.

Reviewer: ```suggestion
> [!NOTE]
> Expected reference values for many of our tests have been calculated on Intel-based CPUs, or CUDA-based GPUs. Precision differences when testing on other hardware may cause certain tests to fail due to calculated values falling outside the configured tolerance limits. These tests may be skipped based on detected hardware (e.g. for [MPS](https://github.com/pytorch/torchtune/blob/ca95345b732d41bab7261d208cd5c860a2f76a5a/tests/test_utils.py#L345).
```

> > Expected reference values for many of our tests have been calculated on Intel-based CPUs, or CUDA-based GPUs. Precision differences when testing on other hardware may cause certain tests to fail due to calculated values falling outside the configured tolerance limits. These tests may be skipped based on detected hardware (e.g. for [MPS](https://github.com/pytorch/torchtune/blob/ca95345b732d41bab7261d208cd5c860a2f76a5a/tests/test_utils.py#L345).

Reviewer: ```suggestion
> Expected reference values for many of our tests have been calculated on Intel-based CPUs, or CUDA-based GPUs. Precision differences when testing on other hardware may cause certain tests to fail due to calculated values falling outside the configured tolerance limits. These tests may be skipped based on detected hardware (e.g. for [MPS](https://github.com/pytorch/torchtune/blob/ca95345b732d41bab7261d208cd5c860a2f76a5a/tests/test_utils.py#L345)).
```

Note that the above flags can be combined with other pytest flags, so e.g. `pytest tests -m integration_test -k 'test_loss'` will run only recipe tests matching the substring `test_loss`.

> It is important to note that outputs computed on MPS may differ from those on CUDA/CPU. As a result, some tests are skipped. Keep this in mind when adding new tests involving calculations.

Reviewer: Similar to my comment above, let's make this a `> [!NOTE]` and instead say that we calculate expected values on CUDA/CPU, so precision differences when testing on other hardware may cause certain tests to fail due to calculated values falling outside the tolerance limits.
